src/functions/includegraphics.js

```diff
@@ -0,0 +1,167 @@
+// @flow
+import defineFunction from "../defineFunction";
+import type {Measurement} from "../units";
+import {calculateSize, validUnit} from "../units";
+import ParseError from "../ParseError";
+import {Img} from "../domTree";
+import mathMLTree from "../mathMLTree";
+import {assertNodeType, checkNodeType} from "../parseNode";
+
+import type {AnyParseNode} from "../parseNode";
+import type {CssStyle} from "../domTree";
+
+// TODO: make it easier for commands to get the macro expanded text
+// from inside an ordgroup argument.
+const stringFromParseGroup = function(textArray: AnyParseNode[]): string {
+    let str = "";
+    for (let i = 0; i < textArray.length; i++) {
+        const textord = checkNodeType(textArray[i], "textord");
+        const spacing = checkNodeType(textArray[i], "spacing");
+        if (textord) {
+            str += textord.text;
+        } else if (spacing) {
+            str += spacing.text;
+        }
+    }
+    return str;
+};
+
+const sizeData = function(str: string): Measurement {
+    if (/^[-+]? *(\d+(\.\d*)?|\.\d+)$/.test(str)) {
+        // str is a number with no unit specified.
+        // default unit is bp, per graphix package.
+        return {number: +str, unit: "bp"};
+    } else {
+        const match = (/([-+]?) *(\d+(?:\.\d*)?|\.\d+) *([a-z]{2})/).exec(str);
+        if (!match) {
+            throw new ParseError("Invalid size: '" + str
+                + "' in \\includegraphics");
+        }
+        const data = {
+            number: +(match[1] + match[2]), // sign + magnitude, cast to number
+            unit: match[3],
+        };
+        if (!validUnit(data)) {
+            throw new ParseError("Invalid unit: '" + data.unit
+             + "' in \\includegraphics.");
+        }
+        return data;
+    }
+};
+
+defineFunction({
+    type: "includegraphics",
+    names: ["\\includegraphics"],
+    props: {
+        numArgs: 1,
+        numOptionalArgs: 1,
+        argTypes: ["text", "text"],
+        allowedInText: false,
+    },
+    handler: ({parser}, args, optArgs) => {
+        let width = {number: 0, unit: "em"};
+        let height = {number: 0.9, unit: "em"};    // sorta character sized.
+        let totalheight = {number: 0, unit: "em"};
+        let alt = "";
+
+        if (optArgs[0]) {
+            const attributeStr = stringFromParseGroup(
+                assertNodeType(optArgs[0], "ordgroup").body);
+            // Parser.js does not parse key/value pairs. We get a string.
+            const attributes = attributeStr.split(",");
+            for (let i = 0; i < attributes.length; i++) {
```

Comment:
My KaTeX work has always been limited. I have avoided several areas, such as font modifications or the inner workings of `makeVList` or `MacroExpander`. `Parser.js` is another in that list. I'm not really very familiar with it.

And adding `key/value` pairs to `Parser.js` would, I think, be a pretty complex addition. Note that my Parse error messages are specific to a type of key, not generic to key/value pairs in general. So if I put the parsing code into `Parser.js` I would have to find a way to pass it a list of what sort of keys and values are contemplated. Then the return value of an argument would be what? An array of ParseNodes? A dictionary? And then I have to write all this into the flow type system?

Nope. That's beyond what I've signed up to do.

Reply:
Fair enough.  I think making it easier to do custom parsing is what we want.  If what's being parsed doesn't include an TeX content then passing a string with the contents of a group to the function handler is the way to go.  I think that's what the intent of the `"raw"` arg type was.  It's a shame the arg types are being ignored.

Reply:
I'll try the `raw` arg type. Also, I think the arg types are being ignored only for functions `\href`, `\url`, and possibly `\verb`. The phantom `url` arg type is not causing errors because it has been used, so far, only by `\href` and `\url`.